### PR TITLE
feat(prepare): Add rev option to base a release on

### LINKED ISSUE #34: Revision option for prepare
Hi! It would be nice to add the `--rev` option for the `prepare` command, is it possible?

### LINKED ISSUE #122: Allow publishing patch releases from a release branch
Right now craft assumes that all your development happens on master.
However sometimes you want to do feature development on master but still keep a previous release branch around to do point releases that don’t have breaking changes.

Right now I have would suggest `releases/0.$N.x`, or `releases/$N.x` as branch names.

Maybe related to #34

### PR DESCRIPTION
**Pull request by BYK**
Fixes #34, fixes #122.

### CODE REVIEW

File: src/commands/prepare.ts [hunk 1]
```diff
@@ -286,10 +274,9 @@ async function checkGitState(
     );
   }
 
-  if (repoStatus.ahead > 0) {
```
Comment (BYK) [inline src/commands/prepare.ts:289]: Removed this check as the way we did it was not reliable anyway: to be able to detect this, we need to run `git fetch` before which we don't.

File: src/commands/prepare.ts [hunk 2]
```diff
@@ -497,30 +463,39 @@ export async function releaseMain(argv: ReleaseOptions): Promise<any> {
...
   // exists.
   const branchName = await createReleaseBranch(
     git,
+    rev,
     newVersion,
     config.releaseBranchPrefix
   );
 
+  // Check & update the changelog
+  await prepareChangelog(
```
Comment (BYK) [inline src/commands/prepare.ts:493]: This step should have come _after_ creating the release branch from the beginning. Old time bug fixed :)

File: src/utils/helpers.ts [hunk 3]
```diff
@@ -1,12 +1,59 @@
+import isCI from 'is-ci';
+import prompts from 'prompts';
+import { logger } from '../logger';
+
+const FALSY_ENV_VALUES = new Set(['', '0', 'false', 'no']);
```
Comment (BYK) [inline src/utils/helpers.ts:5]: Moving of `promptConfirmation` to a helper function triggered a chain of refactors around `hasInput` and `isDryRun`. Apologies for the side quest 🙂

**Review by BYK — COMMENTED** [review]
Will address feedback but need further input on https://github.com/getsentry/craft/pull/223/files#r630370115

File: src/index.ts [hunk 4]
```diff
@@ -37,10 +36,13 @@ function processDryRun<T>(arg: T): T {
...
-  if (arg) {
-    setNoInput(true);
+  // if the user explicitly set the flag on the command line, their choice
+  // should override any previously set env var
+  if (process.argv.indexOf('--no-input') > -1) {
+    process.env.CRAFT_NO_INPUT = String(arg);
   }
-  if (hasNoInput()) {
+
+  if (!hasInput(true)) {
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/index.ts:45]: Decided to defer this to a dedicated PR as there's a larger opportunity to overhaul our yargs usage.

File: src/commands/prepare.ts [hunk 5]
```diff
@@ -497,30 +463,39 @@ export async function releaseMain(argv: ReleaseOptions): Promise<any> {
...
   );
   logger.debug(`Default branch for the repo:`, defaultBranch);
+  const repoStatus = await git.status();
+  const rev = argv.rev || repoStatus.current || defaultBranch;
 
-  // Check that we're in an acceptable state for the release
-  await checkGitState(git, defaultBranch, !argv.noGitChecks);
-
-  // Check whether the version/tag already exists
-  await checkForExistingTag(git, newVersion, !argv.noGitChecks);
```
[the remark this answers is not in the document]
Comment (chadwhitacre) [reply src/commands/prepare.ts:505]: Did we move it to publishing? Does it already exist there?

Probably worth checking at both points, no?

File: src/commands/prepare.ts [hunk 6]
```diff
@@ -497,30 +463,39 @@ export async function releaseMain(argv: ReleaseOptions): Promise<any> {
...
 
-  // Check & update the changelog
-  await prepareChangelog(
-    newVersion,
-    argv.noChangelog ? ChangelogPolicy.None : config.changelogPolicy,
-    config.changelog
-  );
+  logger.info(`Releasing version ${newVersion} from ${rev}`);
+  if (!argv.rev && rev !== defaultBranch) {
+    await promptConfirmation();
```
[the remark this answers is not in the document]
Comment (chadwhitacre) [reply src/commands/prepare.ts:477]: Seeing this:

```
$ ./dist/craft prepare 0.24.0
ℹ info Checking the local repository status...
ℹ info Releasing version 0.24.0 from byk/feat/prepare-from-rev
? Is everything OK? Type "yes" to proceed: ›
```

Maybe add a line like this?

```
$ ./dist/craft prepare 0.24.0
ℹ info Checking the local repository status...
ℹ info Releasing version 0.24.0 from byk/feat/prepare-from-rev
⚠️ warn You're not on your default branch, so I have to ask ...
? Is everything OK? Type "yes" to proceed: ›
```

File: src/utils/helpers.ts [hunk 7]
```diff
@@ -1,12 +1,59 @@
...
+
+let _hasInput: boolean;
+/**
+ * Returns true if user input is allowed. Uses the is-ci module and
+ * the value of CRAFT_NO_INPUT environment variable is checked for
+ * a true-ish value.
+ */
+export function hasInput(clearCache = false): boolean {
+  if (clearCache || _hasInput === undefined) {
+    _hasInput = !(envToBool(process.env.CRAFT_NO_INPUT) ?? isCI);
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/utils/helpers.ts:55]: Don't think so.

**Review by chadwhitacre — APPROVED** [review]

**Pull request merged**